tests/third_party/cupy/manipulation_tests/test_basic.py

```diff
@@ -128,8 +155,83 @@ def test_copyto(self, xp, dtype):
     @testing.numpy_cupy_allclose(accept_error=TypeError)
     def test_copyto_where(self, xp, dtype):
         dst = xp.ones(self.dst_shape, dtype=dtype)
-        mask = (testing.shaped_arange(self.dst_shape, xp, dtype) % 2).astype(
-            xp.bool_
+        # TODO: replace with
+        # mask = (testing.shaped_arange(
+        #     self.dst_shape, xp, dtype) % 2).astype(xp.bool_)
+        # once dpnp.reminder() is fully implemented
+        mask = xp.array(
+            testing.shaped_arange(self.dst_shape, numpy, dtype) % 2,
+            dtype=xp.bool_,
         )
         xp.copyto(dst, self.src, where=mask)
         return dst
+
+
+@pytest.mark.parametrize(
+    "casting", ["no", "equiv", "safe", "same_kind", "unsafe"]
+)
+class TestCopytoFromNumpyScalar:
+    @testing.for_all_dtypes_combination(("dtype1", "dtype2"))
+    @testing.numpy_cupy_allclose(accept_error=TypeError)
+    def test_copyto(self, xp, dtype1, dtype2, casting):
+        if casting == "safe":
+            pytest.skip(
+                "NEP50 doesn't work properly in numpy with casting='safe'"
+            )
+
+        dst = xp.zeros((2, 3, 4), dtype=dtype1)
+        src = numpy.array(1, dtype=dtype2)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", numpy.ComplexWarning)
+            xp.copyto(dst, src, casting)
+        return dst
+
+    @testing.for_all_dtypes()
+    @pytest.mark.parametrize(
+        "make_src",
+        [lambda dtype: numpy.array([1], dtype=dtype), lambda dtype: dtype(1)],
+    )
+    @testing.numpy_cupy_allclose(accept_error=TypeError)
```

Comment:
test_copyto2 does not raise `TypeError` so I think you should remove that

```suggestion
    @testing.numpy_cupy_allclose()
```

Reply:
no, it can't be removed, since some type casting causses exception like:
```
   dpnp
  Traceback (most recent call last):
     File "/home/xantvol/code/dpnp/tests/third_party/cupy/testing/helper.py", line 59, in _call_func
       result = impl(*args, **kw)
     File "/home/xantvol/code/dpnp/tests/third_party/cupy/manipulation_tests/test_basic.py", line 151, in test_copyto
       xp.copyto(dst, self.src)
     File "/home/xantvol/code/dpnp/dpnp/dpnp_iface_manipulation.py", line 357, in copyto
       raise TypeError(
   TypeError: Cannot cast from complex64 to int64 according to the rule same_kind.


   numpy
   Traceback (most recent call last):
     File "/home/xantvol/code/dpnp/tests/third_party/cupy/testing/helper.py", line 59, in _call_func
       result = impl(*args, **kw)
     File "/home/xantvol/code/dpnp/tests/third_party/cupy/manipulation_tests/test_basic.py", line 151, in test_copyto
       xp.copyto(dst, self.src)
     File "<__array_function__ internals>", line 200, in copyto
   TypeError: Cannot cast scalar from dtype('complex128') to dtype('int64') according to the rule 'same_kind'
```
and the test has to secure that both dpnp and numpy raises them for the same pair of types.